File: hahaRU/mainPage/managers/apiManager.py
from typing import Match

from ..exeptions import AuthError, BadRequest
from ..models import *
from datetime import date
from random import randint
from django.core.files.storage import FileSystemStorage
import uuid
from ..moduls.IdList import IdList
from django.http.response import JsonResponse
from base64 import decode
from django.core.files.base import ContentFile
import base64

def objToJSON(data):
    json=data.__dict__
    del json['_state']
    return json

def objsToJSON(data):
    return list(map(objToJSON,data))

def getUser(id):
    user = User.objects.get(pk=int(id));
    if not user:
        raise BadRequest("user with this id doesn't defined")
    user.Password=""
    a=user.__dict__
    del a['_state']
    return a

def getMy(id=0):
    return getUser(id)

def updateUser(data,id):
    keys = data.keys()
    if not "id" in data.keys():
        raise BadRequest("id is required")
    if id != int(data["id"]):
        raise BadRequest("нельзя изменять чужой аккаунт")
    user = User.objects.get(pk=id)
    if not user:
        raise BadRequest("пользователь не найден")
    if "AvatarSrc" in keys:
        user.AvatarSrc = data["AvatarSrc"]
    if "Date" in keys:
        user.Date = date(*map(int, data["Date"].split("-")))
    if "Status" in keys:
        user.Status= data["Status"]
    if "FavoriteJoke" in keys:
        user.FavoriteJoke= data["FavoriteJoke"]
    if "VkLink" in keys:
        user.VkLink = data["VkLink"]
    if "Telegram" in keys:
        user.Telegram = data["Telegram"]
    if "InstaLink" in keys:
        user.InstaLink= data["InstaLink"]
    user.save()
    return {"status":"ok","text":"ok"}


def getContents(data):
    count = 20
    offset = 0
    keys = data.keys()
    if not "type" in keys:
        raise BadRequest("не указан type получаемых данных")
    if "count" in keys:
        count = int(data["count"])
    if "offset" in keys:
        offset = int(data["offset"])
    objs = []
    if data["type"]=="mem":
        objs = objsToJSON(Mem.objects.order_by("-likesCount").all()[offset:offset+count]) 
    elif data["type"]=="funnyWord":
        objs = objsToJSON(FunnyWord.objects.order_by("-likesCount").all()[offset:offset+count])
    elif data["type"]=="video":
        objs = objsToJSON(Video.objects.order_by("-likesCount").all()[offset:offset+count])
    else:
        objs = objsToJSON(Anecdot.objects.order_by("-likesCount").all()[offset:offset+count])
    return {"status":"ok","data":objs}


def getPosts(data):
    orders={
        "new":"-id",
        "old":"id",
        "unpopular":"likesCount"
    }
    count = 20
    offset = 0
    if "count" in data:
        count = int(data["count"])
    if "offset" in data:
        offset = int(data["offset"])
    objs = []
    order = "-likesCount"
    if "order" in data:
        if data["order"] in orders:
            order = orders[data["order"]]
    search=""
    if "search" in data:
        search = data["search"]
    if not "userId" in data:
        objs = objsToJSON(Post.objects.filter(text__contains=search).order_by(order).all()[offset:offset+count])
    else:
        objs = objsToJSON(Post.objects.filter(userId=data["userId"]).order_by("-id").all()[offset:offset+count])
    return {"status":"ok","data":objs}

def getRandomMemImg():
    count = MemPicture.objects.count();
    if not count:
        raise BadRequest("нет изображения")
    id = randint(1,count)
    obj = objToJSON(MemPicture.objects.get(pk=id))
    return {"status":"ok","value":obj}

def getRandomMemText():
    count = MemText.objects.count();
    if not count:
        raise BadRequest("нет Текстов")
    id = randint(1,count)
    obj = objToJSON(MemText.objects.get(pk=id))
    return {"status":"ok","value":obj}

def getRandomVideo():
    count = videoSrc.objects.count();
    if not count:
        raise BadRequest("нет видео")
    id = randint(1,count)
    obj = objToJSON(videoSrc.objects.get(pk=id))
    return {"status":"ok","value":obj}

def sendPost(data,id):
    keys = data.keys()
    if not "text" in keys:
        raise BadRequest("text не указан")
    if not data["text"]:
        raise BadRequest("text не может быть пустым")
    post = Post(text=data["text"],userId=id)
    post.save()
    return {"status":"ok","text":"ok"}


def sendAva(files,id):
    if not len(files):
        raise BadRequest("картинка не найдена")
    availableTypes = {"image/jpeg","image/jpg","image/png"}
    path = ""
    for i in files:
        file = files[i]
        if not file.content_type in availableTypes:
            raise BadRequest("поддерживаются аватарки только в jpg, jpeg, png формате")
        fs = FileSystemStorage()
        names = file.name.split(".");
        type = names[-1];
        name = str(uuid.uuid1())+"."+type;
        filename = fs.save("hahaRU/static/img/avaImgs/"+ name, file)
        user = User.objects.get(pk = id);
        path = "/static/img/avaImgs/"+name
        user.AvatarSrc = path;
        user.save()
    return {"status":"ok","value": path}

def saveMemPic(files):
    if not len(files):
        raise BadRequest("картинка не найдена")
    availableTypes = {"image/jpeg","image/jpg","image/png"}
    path = ""
    for i in files:
        file = files[i]
        if not file.content_type in availableTypes:
            raise BadRequest("поддерживаются аватарки только в jpg, jpeg, png формате")
        fs = FileSystemStorage()
        names = file.name.split(".");
        type = names[-1];
        name = str(uuid.uuid1())+"."+type;
        filename = fs.save("hahaRU/static/img/memImgs/"+ name, file)
        path = name
        memPic = MemPicture(imgSrc= name)
        memPic.save()
    return {"status":"ok","value": path}    

def saveMem(data):
    if not "imgBase64" in data:
        raise BadRequest("не передан imgBase64")
    count = Mem.objects.count();
    path = f"/static/img/memGenerated/img{count}.png"
    with open("hahaRU"+path, "wb") as fs:
        fs.write(decode(str.encode(data["imgBase64"])))
    mem = Mem(imgSrc = path)
    mem.save();
    return {"status" : "ok", "value" : objToJSON(mem)}

def changeContentDisLiked(data,id):
    if not "postId" in data:
        raise BadRequest("Вы не указали поле postId")
    if not "type" in data:
        raise BadRequest("Вы не указали поле type")   
    post = None    
    if data["type"] == "mem":
        post = Mem.objects.get(id=data["postId"])
    elif data["type"] == "funnyWord":
        post = FunnyWord.objects.get(id=data["postId"])
    elif data["type"] == "video":
        post = Video.objects.get(id=data["postId"])
    else:
        post = Anecdot.objects.get(id=data["postId"])
    likes = IdList(post.disLikes)

    if likes.hasId(id):
        likes.removeId(id)
        post.disLikes = likes.toString()
        post.disLikesCount-=1
    else:
        likes.AddId(id)
        post.disLikes = likes.toString()
        post.disLikesCount+=1
    post.save()
    
    return {
        "status" : "ok",
        "text" : "все окей",
        "value" : {"isDisLiked" : 1 if likes.hasId(id) else 0,"disLikesCount" : post.disLikesCount, "id" : post.id }
    }

def changeContentLiked(data,id):
    if not "postId" in data:
        raise BadRequest("Вы не указали поле postId")
    if not "type" in data:
        raise BadRequest("Вы не указали поле type")   
    post = None    
    if data["type"] == "mem":
        post = Mem.objects.get(id=data["postId"])
    elif data["type"] == "funnyWord":
        post = FunnyWord.objects.get(id=data["postId"])
    elif data["type"] == "video":
        post = Video.objects.get(id=data["postId"])
    else:
        post = Anecdot.objects.get(id=data["postId"])
    likes = IdList(post.likes)
    if likes.hasId(id):
        likes.removeId(id)
        post.likes = likes.toString()
        post.likesCount-=1
    else:
        likes.AddId(id)
        post.likes = likes.toString()
        post.likesCount+= 1
    post.save()
    return {
        "status": "ok", 
        "text": "все окей",
        "value": {"isLiked": 1 if likes.hasId(id) else 0, "likesCount": post.likesCount,"id": post.id}
    }



def changeDisLiked(data,id):
    if not "postId" in data:
        raise BadRequest("Вы не указали поле postId")
    post = Post.objects.get(pk= data["postId"])
    likes = IdList(post.disLikes)
    if likes.hasId(id):
        likes.removeId(id)
        post.disLikes = likes.toString()
        post.disLikesCount -= 1
    else:
        likes.AddId(id)
        post.disLikes = likes.toString()
        post.disLikesCount+= 1
    post.save()
    return {
        "status": "ok", 
        "text": "все окей",
        "value": {"isDisLiked": 1 if likes.hasId(id) else 0, "disLikesCount": post.disLikesCount, "id": post.id}
    }


def changeLiked(data,id):
    if not "postId" in data:
        raise BadRequest("Вы не указали поле postId")
    post = Post.objects.get(pk= data["postId"])
    likes = IdList(post.likes)
    if likes.hasId(id):
        likes.removeId(id)
        post.likes = likes.toString()
        post.likesCount-= 1
    else:
        likes.AddId(id)
        post.likes = likes.toString()
        post.likesCount+= 1
    post.save()
    
    return {
        "status": "ok", 
        "text": "все окей",
        "value": {"isLiked": 1 if likes.hasId(id) else 0, "likesCount": post.likesCount,"id": post.id}
    }

File: hahaRU/mainPage/ApiManager.cs
using hahaRU.Lib;
using hahaRU.Models;
using hahaRU.Storage;
using hahaRU.Storage.Entity;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;

namespace hahaRU.Managers
{
    public class ApiManager: IApiManager
    {
        private Context _context;

        public ApiManager(Context context)
        {
            _context = context;
        }

        public object changeContentDisLiked(int postId, object type, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "вы не зарегестрированы" };
            DbSet<Mem> db;
            IContent post;
            switch (type)
            {
                case "mem":
                    post = _context.Mems.Where(post => post.Id == postId).First();
                    break;
                case "funnyWord":
                    post = _context.FunnyWords.Where(post => post.Id == postId).First();
                    break;
                case "video":
                    post = _context.Videos.Where(post => post.Id == postId).First();
                    break;
                default:
                    post = _context.Anecdots.Where(post => post.Id == postId).First();
                    break;
            }
            if (post == null) return new JsonStatus() { status = "error", text = "post не существует" };
            IdList likes = new IdList(post.Dislikes);
            if (likes.hasId((int)id))
            {
                likes.removeId((int)id);
                post.Dislikes = likes.toString();
                post.DislikesCount--;
            }
            else
            {
                likes.AddId((int)id);
                post.Dislikes = likes.toString();
                post.DislikesCount++;
            }
            _context.SaveChanges();
            return new JsonStatus()
            {
                status = "ok",
                text = "все окей",
                value = new ContentData()
                {
                    IsDisLiked = likes.hasId((int)id) ? 1 : 0,
                    DislikesCount = post.DislikesCount,
                    Id = post.Id
                }
            };
        }

        public object changeContentLiked(int postId, string type, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "вы не зарегестрированы" };
            IContent post;
            switch (type)
            {
                case "mem":
                    post = _context.Mems.Where(post => post.Id == postId).First();
                    break;
                case "funnyWord":
                    post = _context.FunnyWords.Where(post => post.Id == postId).First();
                    break;
                case "video":
                    post = _context.Videos.Where(post => post.Id == postId).First();
                    break;
                default:
                    post = _context.Anecdots.Where(post => post.Id == postId).First();
                    break;
            }
            if (post == null) return new JsonStatus() { status = "error", text = "post не существует" };
            IdList likes = new IdList(post.Likes);
            if (likes.hasId((int)id))
            {
                likes.removeId((int)id);
                post.Likes = likes.toString();
                post.LikesCount--;
            }
            else
            {
                likes.AddId((int)id);
                post.Likes = likes.toString();
                post.LikesCount++;
            }
            _context.SaveChanges();
            return new JsonStatus()
            {
                status = "ok",
                text = "все окей",
                value = new ContentData()
                {
                    IsLiked = likes.hasId((int)id) ? 1 : 0,
                    LikesCount = post.LikesCount,
                    Id = post.Id
                }
            };
        }

        public object changeDisLiked(int postId, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "вы не зарегестрированы" };
            List<Post> posts = _context.Posts.Where(post => post.Id == postId).ToList();
            if (posts.Count == 0) return new JsonStatus() { status = "error", text = "post не существует" };
            Post post = posts[0];
            IdList likes = new IdList(post.Dislikes);
            if (likes.hasId((int)id))
            {
                likes.removeId((int)id);
                post.Dislikes = likes.toString();
                post.DislikesCount--;
            }
            else
            {
                likes.AddId((int)id);
                post.Dislikes = likes.toString();
                post.DislikesCount++;
            }
            _context.SaveChanges();
            return new JsonStatus() { status = "ok", 
                text = "все окей",
                value = new PostData()
                {
                    IsDisLiked = likes.hasId((int)id) ? 1 : 0,
                    DislikesCount = post.DislikesCount,
                    Id = post.Id
                }
            };
        }

        public object changeLiked(int postId, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "вы не зарегестрированы" };
            List<Post> posts = _context.Posts.Where(post => post.Id == postId).ToList();
            if(posts.Count==0) return new JsonStatus() { status = "error", text = "post не существует" };
            Post post = posts[0];
            IdList likes = new IdList(post.Likes);
            if(likes.hasId((int)id))
            {
                likes.removeId((int)id);
                post.Likes = likes.toString();
                post.LikesCount--;
            }
            else
            {
                likes.AddId((int)id);
                post.Likes = likes.toString();
                post.LikesCount++;
            }
            _context.SaveChanges();
            return new JsonStatus() { 
                status = "ok", 
                text = "все окей", 
                value = new PostData() {
                    IsLiked= likes.hasId((int)id)?1:0,
                    LikesCount= post.LikesCount,
                    Id=post.Id
                } 
            };
        }

        public object getContents(getPostReq data, HttpContext httpContext)
        {
            int count = data.Count ?? 20;
            int offset = data.Offset ?? 0;
            int? id = httpContext.Session.GetInt32("id");
            id = (id == null) ? 0 : id;
            List<IContent> posts;
            switch (data.type)
            {
                case "mem":
                    posts = _context.Mems.OrderByDescending(x => x.LikesCount).Skip(offset).Take(count).ToList().Cast<IContent>().ToList();
                    break;
                case "funnyWord":
                    posts = _context.FunnyWords.OrderByDescending(x => x.LikesCount).Skip(offset).Take(count).ToList().Cast<IContent>().ToList();
                    break;
                case "video":
                    posts = _context.Videos.OrderByDescending(x => x.LikesCount).Skip(offset).Take(count).ToList().Cast<IContent>().ToList();
                    break;
                default:
                    posts = _context.Anecdots.OrderByDescending(x => x.LikesCount).Skip(offset).Take(count).ToList().Cast<IContent>().ToList();
                    break;
            }
            List<ContentData> Contents = new List<ContentData>();
            foreach (IContent post in posts)
            {
                Contents.Add(new ContentData()
                {
                    Id = post.Id,
                    Text = post.Text,
                    Date = post.Date,
                    ImgSrc = post.ImgSrc,
                    VideoSrc=post.VideoSrc,
                    IsLiked = (new IdList(post.Likes)).hasId((int)id) ? 1 : 0,
                    IsDisLiked = (new IdList(post.Dislikes)).hasId((int)id) ? 1 : 0,
                    LikesCount = post.LikesCount,
                    DislikesCount = post.DislikesCount
                });
            }
            return Contents;
        }

        public object getPosts(getPostReq data, HttpContext httpContext)
        {
            int count= data.Count ?? 20;
            int offset = data.Offset ?? 0;
            int? id = httpContext.Session.GetInt32("id");
            id = (id==null)?0:id;
            List<Post> posts;
            if (data.UserId == null)
                posts= _context.Posts.OrderByDescending(x => x.LikesCount).Skip(offset).Take(count).ToList();
            else
                posts = _context.Posts.Where(e => e.UserId == data.UserId).OrderByDescending(x => x.Id).Skip(offset).Take(count).ToList();

            List<PostData> Posts = new List<PostData>();
            foreach(Post post in posts)
            {
                Posts.Add(new PostData()
                {
                    Id = post.Id,
                    Text = post.Text,
                    Date = post.Date,
                    UserId = post.UserId,
                    ImgSrc = post.ImgSrc,
                    IsLiked = (new IdList(post.Likes)).hasId((int)id) ? 1: 0,
                    IsDisLiked = (new IdList(post.Dislikes)).hasId((int)id) ? 1 : 0,
                    LikesCount=post.LikesCount,
                    DislikesCount=post.DislikesCount
                });
            }
            return Posts;
        }

        public object getRundomAnecdot()
        {
            return new object();
        }

        public object getRundomMemImg()
        {
            int count = _context.memPictures.Count();
            if (count == 0) return new JsonStatus() { status = "error", text = "нет изображений" };
            Random rnd = new Random();
            int id = rnd.Next(1, count+1);
            memPictures mem = _context.memPictures.Where(mem => mem.Id == id).First();
            return new JsonStatus() { status = "ok", value = mem };
        }

        public object getRundomMemText()
        {
            int count = _context.memTexts.Count();
            if (count == 0) return new JsonStatus() { status = "error", text = "нет текстов" };
            Random rnd = new Random();
            int id = rnd.Next(1, count+1);
            memText mem = _context.memTexts.Where(mem => mem.Id == id).First();
            return new JsonStatus() { status = "ok", value=mem };
        }

        public object getRundomVideo()
        {
            int count = _context.VideoSrcs.Count();
            if (count == 0) return new JsonStatus() { status = "error", text = "нет видео" };
            Random rnd = new Random();
            int id = rnd.Next(1, count);
            VideoSrc video = _context.VideoSrcs.Where(mem => mem.Id == id).First();
            return new JsonStatus() { status = "ok", value = video };
        }

        public string getUser(int id)
        {
            var users = _context.Users.Where(e => e.Id == id).ToList();
            if (users.Count == 0) return "{}";
            users[0].Password = null;
            return JsonSerializer.Serialize(users[0]);
        }

        public string getUser(HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return "{}";
            var users = _context.Users.Where(e => e.Id == id).ToList();
            if (users.Count == 0) return "{}";
            users[0].Password = null;
            return JsonSerializer.Serialize(users[0]);
        }

        public object getVideos(getPostReq data, HttpContext httpContext)
        {
            throw new NotImplementedException();
        }
        public object saveAva(IFormFileCollection files, string webRootPath, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "Вы не авторизовались" };
            string path = "";
            if(files.Count==0) return new JsonStatus() { status = "error", text = "Картинка не найдена" };
            var file = files[0];
            path = "/img/avaImgs/" + file.FileName;
                using (var fileStream = new FileStream(webRootPath + path, FileMode.Create))
                {
                    file.CopyTo(fileStream);
                }
                memPictures pic = new memPictures() { ImgSrc = file.FileName };
            User user = _context.Users.Single(user => user.Id == (int)id);
            if(user==null) return new JsonStatus() { status = "error", text = "Вы не авторизовались 2" };
            user.AvatarSrc = path;
            _context.SaveChanges();
            return new JsonStatus() { status = "ok", value = path };
        }

        public object saveMem(IFormFileCollection files, string webRootPath)
        {
            foreach (var file in files)
            {
                string path = "/img/memGenerated/" + file.FileName;
                using (var fileStream = new FileStream(webRootPath + path, FileMode.Create))
                {
                    file.CopyTo(fileStream);
                }
                Mem mem = new Mem() { ImgSrc = path };
                _context.Mems.Add(mem);
            }
            _context.SaveChanges();
            return new object();
        }

        public object saveMem(string imgBase64, string webRootPath)
        {
            int count = _context.Mems.Count();
            string path = $"/img/memGenerated/img{count}.png";
            File.WriteAllBytes(webRootPath + path, Convert.FromBase64String(imgBase64));
            Mem mem = new Mem() { ImgSrc = path };
            _context.Mems.Add(mem);
            _context.SaveChanges();
            return new JsonStatus() { status = "ok",value=mem };
        }

        public object saveMemPic(IFormFileCollection files, string webRootPath)
        {
            string path="";
            foreach (var file in files)
            {
                path = "/img/memImgs/" + file.FileName;
                using (var fileStream = new FileStream(webRootPath + path, FileMode.Create))
                {
                    file.CopyTo(fileStream);
                }
                memPictures pic = new memPictures() { ImgSrc = file.FileName };
                _context.memPictures.Add(pic);
                _context.SaveChanges();
            }
           return new JsonStatus() { status = "ok", value = path };
        }
            
        

        public object sendPost(Post post, HttpContext httpContext)
        {
            int? id = httpContext.Session.GetInt32("id");
            if (id == null) return new JsonStatus() { status = "error", text = "Вы не авторизовались" };
            if (post.Text == "") return new JsonStatus() { status = "error", text = "Пост не может быть без текста" };
            if (post.Text == null) return new JsonStatus() { status = "error", text = "текст не найден" };
            Post post1 = new Post()
            {
                Text = post.Text,
                UserId = (int)id,
                Date = post.Date
            };
            _context.Posts.Add(post1);
            _context.SaveChanges();
            return new JsonStatus() { status="ok",text="ok"};
        }

        public string updateUser(User user, HttpContext httpContext)
        {
            if (user.Id==0) return "Id is not find";
            int? id = httpContext.Session.GetInt32("id");
            if(id==null) return "Your doesn't auth";
            if(id!=user.Id) return "нельзя изменять чужой аккаунт";
            User userNew = _context.Users
                .Where(c => c.Id == id)
                .First();
            if (userNew == null) return "how";
            if (user.AvatarSrc != null) userNew.AvatarSrc = user.AvatarSrc;
            if (user.Date != null) userNew.Date = user.Date;
            if (user.Status != null) userNew.Status = user.Status;
            if (user.FavoriteJoke != null) userNew.FavoriteJoke = user.FavoriteJoke;
            if (user.VkLink != null) userNew.VkLink = user.VkLink;
            if (user.Telegram != null) userNew.Telegram = user.Telegram;
            if (user.InstaLink != null) userNew.InstaLink = user.InstaLink;
            _context.SaveChanges();
            return "ok";
        }

        
    }
}
